modules/node/src/NodeAccessControlHandler.php

```diff
@@ -27,17 +28,49 @@ class NodeAccessControlHandler extends EntityAccessControlHandler implements Nod
    */
   protected $grantStorage;
 
+  /**
+   * Node entity storage.
+   *
+   * @var \Drupal\node\NodeStorageInterface
+   */
+  protected $nodeStorage;
+
+  /**
+   * Map of revision operations.
+   *
+   * Keys contain revision operations, where values are an array containing the
+   * permission operation and supplemental entity operation.
+   *
+   * Permission operation is used to build the required permission, e.g.
+   * 'permissionOperation all revisions', 'permissionOperation type revisions'.
+   *
+   * Supplemental entity operation is used to determine access, for
+   * 'delete revision' operation, an account must also have access to 'delete'
+   * operation on an entity.
+   *
+   * @internal Only to be used internally by this class. Consider as protected.
+   */
+  const REVISION_OPERATION_MAP = [
+    'view all revisions' => ['view', 'view'],
+    'view revision' => ['view', 'view'],
+    'revert' => ['revert', 'update'],
+    'delete revision' => ['delete', 'delete'],
+  ];
+
   /**
    * Constructs a NodeAccessControlHandler object.
    *
    * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
    *   The entity type definition.
    * @param \Drupal\node\NodeGrantDatabaseStorageInterface $grant_storage
    *   The node grant storage.
+   * @param \Drupal\Core\Entity\EntityStorageInterface|null $nodeStorage
+   *   Node entity storage.
    */
-  public function __construct(EntityTypeInterface $entity_type, NodeGrantDatabaseStorageInterface $grant_storage) {
+  public function __construct(EntityTypeInterface $entity_type, NodeGrantDatabaseStorageInterface $grant_storage, EntityStorageInterface $nodeStorage = NULL) {
     parent::__construct($entity_type);
     $this->grantStorage = $grant_storage;
+    $this->nodeStorage = isset($nodeStorage) ? $nodeStorage : \Drupal::entityTypeManager()->getStorage('node');
```

Comment:
we should `trigger_error` here too if the arg is missing so folks know and we can enforce it in d9